aipcli/command.go

```diff
@@ -0,0 +1,259 @@
+) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   methodUse(method),
+		Short: initialUpperCase(trimComment(comments[method.FullName()])),
+		Long:  comments[method.FullName()],
+	}
+	fromFile := cmd.Flags().StringP("from-file", "f", "", "path to a JSON file containing the request payload")
+	_ = cmd.MarkFlagFilename("from-file", "json")
+	setFlags(comments, cmd.Flags(), nil, in.ProtoReflect().Descriptor(), in.ProtoReflect)
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		if cmd.Flags().Changed("from-file") {
+			data, err := os.ReadFile(*fromFile)
+			if err != nil {
+				return err
+			}
+			if err := protojson.Unmarshal(data, in); err != nil {
+				return err
+			}
+		}
+		conn, err := Dial(cmd.Context())
+		if err != nil {
+			return err
+		}
+		LogRequest(cmd.Context(), in)
+		if err := conn.Invoke(cmd.Context(), methodURI(method), in, out); err != nil {
+			LogError(cmd.Context(), err)
+			os.Exit(1)
+		}
+		LogResponse(cmd.Context(), out)
+		return nil
+	}
+	return cmd
+}
+
+func serviceUse(service protoreflect.ServiceDescriptor) string {
+	result := string(service.Name())
+	result = strings.TrimSuffix(result, "Service")
+	result = strcase.KebabCase(result)
+	return result
+}
+
+func methodUse(method protoreflect.MethodDescriptor) string {
+	result := string(method.Name())
+	result = strcase.KebabCase(result)
+	return result
+}
+
+func methodURI(method protoreflect.MethodDescriptor) string {
+	return "/" +
+		string(method.Parent().(protoreflect.ServiceDescriptor).FullName()) +
+		"/" + string(method.Name())
+}
+
+func setFlags(
+	comments map[protoreflect.FullName]string,
+	flags *pflag.FlagSet,
+	parentFields []protoreflect.FieldDescriptor,
+	msg protoreflect.MessageDescriptor,
+	mutable func() protoreflect.Message,
+) {
+	for i := 0; i < msg.Fields().Len(); i++ {
+		field := msg.Fields().Get(i)
+		switch field.Kind() {
+		case protoreflect.MessageKind:
+			switch field.Message().FullName() {
+			case "google.protobuf.Duration":
+				if field.IsList() {
+					// TODO: Implement support for repeated durations.
+				} else {
+					flags.AddFlag(&pflag.Flag{
+						Name:  flagName(field, parentFields),
+						Usage: flagUsage(comments[field.FullName()]),
+						Value: durationValue{mutable: mutable, field: field},
+					})
+				}
+			case "google.protobuf.Timestamp":
+				if field.IsList() {
+					// TODO: Implement support for repeated timestamps.
+				} else {
+					flags.AddFlag(&pflag.Flag{
+						Name:  flagName(field, parentFields),
+						Usage: flagUsage(comments[field.FullName()]),
+						Value: timestampValue{mutable: mutable, field: field},
+					})
+				}
+			case "google.protobuf.FieldMask":
+				if field.IsList() {
+					// Repeated field masks is intentionally not supported.
+				} else {
+					flags.AddFlag(&pflag.Flag{
+						Name:  flagName(field, parentFields),
+						Usage: flagUsage(comments[field.FullName()]),
+						Value: fieldMaskValue{mutable: mutable, field: field},
+					})
+				}
+			default:
```

Comment:
is this for nested fields?

Reply:
yeah! that's how we get `--shipment.name` etc flags in e.g. Update methods

Reply:
ah! cool!